File: Python/Nonlineq/expNewton.py

# solution for min_(A,E,T0) || rhs - A*exp(E*(1./(t-T0))) ||
import numpy as np
import numpy.linalg as ln #Matrix solvers/eigenvalues
import matplotlib.pyplot as plt #Plotting
from scipy.stats import norm

#genererar högerleden
A = 1; E = 1; T0 = 200
t = np.linspace(210,270,50)

#genererar data punkter
b = A*np.exp(E*(1./(t-T0)))

#genererar observationer  med random brus
random1=norm.ppf(np.random.rand(len(b))) #random med normal distrobution
brus = 0.01
rhs =  b + brus*random1



def func(x):
    return x[0]*np.exp(x[1]*(1/(t-x[2])))  #här är python / samma som matlab ./

def residual(x):
    return rhs - x[0]*np.exp(x[1]*(1/(t-x[2])))   

B=1

def derfunA(x):
    return -1*np.exp(x[1]*(1/(t-x[2])))

def derfunE(x):
    return B*np.exp(x[1]*(1/(t-x[2])))

def derfunT0(x):
    return x[0]*np.exp(x[1]*(1/(t-x[2])))


# Newtons metod

iteration_newton =[]
func_newton = []


#Initialiseringen
tolerance=10**(-15)
z0 = np.array([1,1,201.0]) #starta som fult array
z = np.array([1,0,201.0]) #decimal behövs för att array inte auto till int
zold = z.copy() 

k=1
iteration_newton.append(1)
func_newton = 0


numerator =  residual(z)
 
B= -z[0]*(1/(t-z[2]));

denominator0 = [derfunA(z)] #Extra hawk för P-inv
denominator1 = [derfunE(z)]
denominator2 = [derfunT0(z)]

#Moore-Penrose pseudoinverse beräknas här
z[0] = z[0] - np.dot(numerator, ln.pinv(denominator0))[0]
z[1] = z[1] - np.dot(numerator, ln.pinv(denominator1))[0]
z[2] = z[2] - np.dot(numerator, ln.pinv(denominator2))[0]

while ln.norm(abs(z-zold))>tolerance:
    zold=z.copy() #copy för attt undvika tidigt stop
    
    numerator =  residual(z)

    B= -z[0]*(1/(t-z[2]))

    denominator0 = [derfunA(z)]
    denominator1 = [derfunE(z)]
    denominator2 = [derfunT0(z)]

    #Moore-Penrose pseudoinverse beräknas här
    z[0] = z[0] - np.matmul(numerator, ln.pinv(denominator0))[0]
    z[1] = z[1] - np.matmul(numerator, ln.pinv(denominator1))[0]
    z[2] = z[2] - np.matmul(numerator, ln.pinv(denominator2))[0] #z updateras
    
    k=k+1
    iteration_newton.append(k)
    if k>1000:
        break 
    
    func_newton = func(z)


plt.plot(t,func_newton, 'c-',label="Newtons metod")

plt.plot(t,b, 'r--',label="exact solution")
  
plt.plot(t,rhs, 'b--',label="noisy data")
  
plt.xlabel('t')
plt.ylabel('funktion')
plt.title("Inital guess =" f'{z0[0]}' ",E=" f'{z0[1]}' ",T0="f'{z0[2]}')
plt.legend()


print(iteration_newton)

File: MATLAB/Nonlineq/expNewton.m
% solution for min_(A,E,T0) || rhs - A*exp(E*(1./(t-T0))) ||

close all;

%genererar högerleden
A=1;
E=1;
T0 =200;

%  genererar data punkter
t = linspace(210,270,50);

% genererar exakt funktion i data punkter

b = A*exp(E*(1./(t-T0)));


% genererar observationer  med random brus

brus = 0.01;
rhs =  b + brus*randn(size(t));



func = @(x)x(1)*exp(x(2)*(1./(t-x(3))));
residual= @(x)rhs - x(1)*exp(x(2)*(1./(t-x(3))));

B =1;
C=1;

derfunA = @(x)-exp(x(2)*(1./(t-x(3))));
derfunE = @(x)B*exp(x(2)*(1./(t-x(3))));
%derfunT0 = @(x)x(1)*exp(x(2)*(1./(t-x(3))));
derfunT0 = @(x)C*exp(x(2)*(1./(t-x(3))));

% Newtons metod



z=0;
iteration_newton = 0;
func_newton = 0;


%Initialiseringen
%z = [1,1,201];
z = [1,1,190];

z0 = [0,0,150];

k=2;

iteration_newton(1) = 1;
func_newton = residual(z);
numerator =  residual(z);

 
denominator1 = derfunA(z);

B= -z(1)*(1./(t-z(3)));

C = -z(1)*z(2)*(1./(t - z(3)).^2);

denominator2 = derfunE(z);
denominator3 = derfunT0(z);

znew(1) = z(1) - numerator/denominator1;
znew(2) = z(2) - numerator/denominator2;
znew(3) = z(3) - numerator/denominator3;

 zold = z; 
% Main Newton's iterations  
while abs(znew - zold) > tolerance

numerator =  residual(z);

denominator1 = derfunA(z);

B= -z(1)*(1./(t-z(3)));

denominator2 = derfunE(z);
denominator3 = derfunT0(z);

znew(1) = z(1) - numerator/denominator1;
znew(2) = z(2) - numerator/denominator2;
znew(3) = z(3) - numerator/denominator3;


iteration_newton(k) = k;

func_newton =  func(z);

k = k+1;


zold = z;
z= znew;

if k > 10000
    break
end
  end

  plot(t,func_newton, '- c','LineWidth',2)
hold on

  plot(t,b, '-- r','LineWidth',2)
  
  plot(t,rhs, '-- b','LineWidth',2)
  
  
  xlabel('t')
  ylabel('funktion')
  legend('Newtons metod','exact solution','noisy data')
title([' Initial guess A=',num2str(z0(1)),', E= ',num2str(z0(2)),', T0= ',num2str(z0(3))])

 iteration_newton
